test.js

```diff
@@ -98,10 +98,15 @@ test('test simple WriteStream', function (t, ctx, done) {
   ws.on('error', function (err) {
     t.notOk(err, 'no error')
   })
-  ws.on('close', ctx.verify.bind(ctx, ws, done))
+
+  var next = after(2, ctx.verify.bind(ctx, ws, done))
+  ctx.db.once('batch', next.bind(null, null))
```

Comment:
Using the `'batch'` event from `levelup` here. We need to wait until the batch operation has ended successfully before we can verify that the data has been written (or deleted) correctly.

Reply:
Doesn't the write stream's `close` event wait for the batch to complete?

Reply:
If that's the case, then the issue is in the code, and this PR will hide it. Correct me if I'm wrong.

Reply:
I think this code is the problem:

https://github.com/Level/level-ws/blob/35793fc4b6dfe1c3959e3d4c8d1897be82207dda/level-ws.js#L64-L70

It would be more correct to flush before `finish` (and this is something that #35 can help with), and only then emit `close`.

Reply:
> Doesn't the write stream's close event wait for the batch to complete?

That's exactly the problem.

> If that's the case, then the issue is in the code, and this PR will hide it. Correct me if I'm wrong.

You are probably right about this. Glad this PR brought it up at least 😄

Reply:
Hmm, that code looks like it should work, since it calls `._flush()` and emits `'close'` once back from flush.

Reply:
Aah _now_ I see why @mafintosh 's module is good for this

Reply:
We could also use [`_final`](https://nodejs.org/api/stream.html#stream_writable_final_callback), which is like `_flush` for writable streams.

Reply:
E.g.:

```js
WriteStream.prototype._final = function (callback) {
  if (this._buffer && this._buffer.length) { 
     return this._flush(callback) 
  }

  callback()
}
```

Reply:
Up for making a PR or want me to try it out?

Reply:
Not the following?

```js
WriteStream.prototype._final = function (callback) {
  if (this._buffer && this._buffer.length) { 
     return this._flush(callback) 
  }

  process.nextTick(callback)
}
```

Reply:
Good question. I *assume* that's not necessary. The [`_final` method itself is already called in the next tick](https://github.com/nodejs/readable-stream/blob/862ab4f1f8e862602251afed4fe5a146fb98b5a3/lib/_stream_writable.js#L606), so another `process.nextTick` would be redundant I think.

Reply:
I think we should make sure that `readable-stream` is used before we do anything else. I'll get that done.

Reply:
I tried the following, but it also fails after a while:

```js
WriteStream.prototype._final = function (callback) {
  if (this._buffer && this._buffer.length) { 
     return this._flush(callback) 
  }

  this.writable = false
  this.emit('close')

  callback()
}
```

(and removed the `this.on('finish')` handler)

Reply:
I think this emits `close` before `finish`, if there's nothing to flush. Close should be emitted after finish (or not at all. I'm not sure this event has any meaning here, at least not more than `finish`).

Maybe try keeping `this.on('finish')`, emit the `close` event from there, and make sure this is the only code path - except for `destroy()` - that emits `close`.

Btw, we could also refactor `destroy` into [`_destroy`](https://nodejs.org/api/stream.html#stream_writable_destroy_err_callback). There's probably more custom logic we can remove, for which better primitives (`_final`, `_destroy`, perhaps `_writev`) have since been introduced in streams.

Wish I could help more but busy this week.